src/transformers/models/seamless_m4t/modeling_seamless_m4t.py

```diff
@@ -0,0 +1,4500 @@
+                            return module(*inputs, output_attentions)
+
+                        return custom_forward
+
+                    layer_outputs = torch.utils.checkpoint.checkpoint(
+                        create_custom_forward(encoder_layer),
+                        hidden_states,
+                        attention_mask,
+                        (head_mask[idx] if head_mask is not None else None),
+                    )
+                else:
+                    layer_outputs = encoder_layer(
+                        hidden_states,
+                        attention_mask,
+                        layer_head_mask=(head_mask[idx] if head_mask is not None else None),
+                        output_attentions=output_attentions,
+                    )
+
+                hidden_states = layer_outputs[0]
+
+            if output_attentions:
+                all_attentions = all_attentions + (layer_outputs[1],)
+
+        hidden_states = self.layer_norm(hidden_states)
+
+        if output_hidden_states:
+            encoder_states = encoder_states + (hidden_states,)
+
+        if not return_dict:
+            return tuple(v for v in [hidden_states, encoder_states, all_attentions] if v is not None)
+        return BaseModelOutput(
+            last_hidden_state=hidden_states, hidden_states=encoder_states, attentions=all_attentions
+        )
+
+
+class SeamlessM4TDecoder(SeamlessM4TPreTrainedModel):
+    """
+    Transformer decoder consisting of *config.decoder_layers* layers. Each layer is a [`SeamlessM4TDecoderLayer`]
+
+    Args:
+        config: (`SeamlessM4TConfig`)
+        embed_tokens (`nn.Embedding`, *optional*): output embedding
+        is_t2u_decoder (`bool`, *optional*, defaults to `False`): indicates if it belongs to the text-to-units model
+    """
+
+    def __init__(
+        self,
+        config: SeamlessM4TConfig,
+        embed_tokens: Optional[nn.Embedding] = None,
+        is_t2u_decoder: Optional[bool] = False,
+    ):
+        super().__init__(config)
+        self.dropout = config.dropout
+        self.layerdrop = config.decoder_layerdrop
+        self.padding_idx = config.t2u_pad_token_id if is_t2u_decoder else config.pad_token_id
+        self.vocab_size = config.unit_vocab_size if is_t2u_decoder else config.vocab_size
+        self.max_target_positions = (
+            config.t2u_max_position_embeddings if is_t2u_decoder else config.max_position_embeddings
+        )
+        self.embed_scale = math.sqrt(config.hidden_size) if config.scale_embedding else 1.0
+        decoder_layers = config.t2u_decoder_layers if is_t2u_decoder else config.decoder_layers
+        decoder_attention_heads = (
+            config.t2u_decoder_attention_heads if is_t2u_decoder else config.decoder_attention_heads
+        )
+        decoder_ffn_dim = config.t2u_decoder_ffn_dim if is_t2u_decoder else config.decoder_ffn_dim
+
+        if embed_tokens is not None:
+            # if embed_tokens defined, use its shape instead
```

Comment:
Yes exactly! It's not exactly implied since there's also the `unit_vocab_size` parameter that is specified in the config and that is coherent with this code! Feel free to ask for a refactoring if necessary